crates/eips/src/eip7702/auth_list.rs

```diff
@@ -0,0 +1,119 @@
+use core::ops::Deref;
+
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+use alloy_primitives::{keccak256, Address, ChainId, Signature, SignatureError, B256};
+use alloy_rlp::{BufMut, Decodable, Encodable, Header, RlpEncodable};
+
+use super::constants::MAGIC;
+
+/// An EIP-7702 authorization list.
+pub type AuthorizationList = Vec<Authorization>;
+
+/// An EIP-7702 authorization.
+#[derive(Debug, Clone, RlpEncodable)]
+pub struct Authorization {
+    chain_id: ChainId,
+    address: Address,
+    nonce: OptionalNonce,
+    signature: Signature,
+}
+
+impl Authorization {
+    /// Get the `chain_id` for the authorization.
+    ///
+    /// # Note
+    ///
+    /// Implementers should check that this matches the current `chain_id` *or* is 0.
+    pub const fn chain_id(&self) -> ChainId {
+        self.chain_id
+    }
+
+    /// Get the `address` for the authorization.
+    pub const fn address(&self) -> &Address {
+        &self.address
+    }
+
+    /// Get the `nonce` for the authorization.
+    ///
+    /// # Note
+    ///
+    /// If this is `Some`, implementers should check that the nonce of the authority (see
+    /// [`Self::recover_authority`]) is equal to this nonce.
+    pub fn nonce(&self) -> Option<u64> {
+        *self.nonce
+    }
+
+    /// Get the `signature` for the authorization.
+    pub const fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    /// Computes the authority prehash used to recover the authority from an authorization list
+    /// item.
+    ///
+    /// The authority prehash is `keccak(MAGIC || rlp([chain_id, [nonce], address]))`
+    #[inline]
+    #[cfg(feature = "k256")]
+    fn authority_prehash(&self) -> B256 {
+        #[derive(RlpEncodable)]
+        struct Auth {
+            chain_id: ChainId,
+            nonce: OptionalNonce,
```

Comment:
is this the correct order?

this is a bit weird, I would expect that the optional stuff is the last thing

Reply:
yes, it is correct.

>At the start of executing the transaction, for each [chain_id, address, [nonce], y_parity, r, s] tuple:
>
>authority = ecrecover(keccak(MAGIC || rlp([chain_id, [nonce], address])), y_parity, r, s]

Reply:
jfc